src/Conversation/conversationManager.ts

```diff
@@ -0,0 +1,183 @@
+import { Conversation } from "../ChatModel/ChatModel";
+import { ExtensionContext } from "vscode";
+import { ChatMessage } from "../ChatModel/ChatModel";
+import { CONTEXT_INSTRUCTION } from "../commands/constants";
+
+export class ConversationManager {
+  // TODO: Use constants where needed
+  private readonly context: ExtensionContext;
+
+  constructor(context: ExtensionContext) {
+    this.context = context;
+  }
+
+  get conversations(): Conversation[] {
+    return (
+      this.context.workspaceState.get<Conversation[]>("conversations") || []
+    );
+  }
+
+  set conversations(value: Conversation[]) {
+    const ids = value.map((conversation) => conversation.id);
+    if (new Set(ids).size !== ids.length) {
+      throw new Error("Duplicate ids");
+    }
+    this.context.workspaceState.update("conversations", value);
+  }
+
+  get conversationIds(): number[] {
+    return this.context.workspaceState.get<number[]>("conversationIds") || [];
+  }
+
+  set conversationIds(value: number[]) {
+    if (new Set(value).size !== value.length) {
+      throw new Error("Duplicate ids");
+    }
+    this.context.workspaceState.update("conversationIds", value);
+  }
+
+  get activeConversationId(): number | null {
+    return (
+      this.context.workspaceState.get<number>("activeConversationId") || null
+    );
+  }
+
+  set activeConversationId(value: number | null) {
+    if (value && !this.conversationIds.includes(value)) {
+      throw new Error("Conversation ID does not exist");
+    }
+    this.context.workspaceState.update("activeConversationId", value);
+  }
+
+  getConversation(id: number): Conversation | undefined {
+    return this.conversations.find((conversation) => conversation.id === id);
+  }
+
+  getActiveConversation(): Conversation | undefined {
+    if (!this.activeConversationId) {
+      return undefined;
+    }
+    return this.getConversation(this.activeConversationId);
+  }
+
+  /**
+   *
+   * @returns An array of conversation previews, each containing the conversation id and name.
+   */
+  getConversationPreviews(): {
+    id: number;
+    name: string;
+  }[] {
+    return this.conversations.map((conversation) => {
+      return {
+        id: conversation.id,
+        name: conversation.name,
+      };
+    });
+  }
+
+  /**
+   * Adds a conversation to the conversation history and sets it as the active conversation.
+   * Throws an error if a conversation with the same id already exists.
+   *
+   * @param name Conversation name
+   * @param messages Conversation messages
+   * @returns The created conversation
+   */
+  startConversation(name: string, messages?: ChatMessage[]): Conversation {
+    const conversation = this.createConversation(name, messages);
+    this.conversations = [...this.conversations, conversation];
```

Comment:
Is there a reason that you're copying `this.conversations` instead of just doing `this.conversations.push(conversation)`?

Reply:
Nope. Made the change!